### PR TITLE
Add recursion limit to fromBinary and fromJson

### PR DESCRIPTION
**Pull request by timostamm**
Decoding does not currently limit recursion depth, so deeply-nested input can trivially exhaust the call stack.

This adds a `recursionLimit` option to both binary and ProtoJSON parsing:

```ts
  /**
   * The maximum depth of nested messages to parse. If a message nests deeper
   * than this limit, parsing fails with an error instead of exhausting the
   * call stack. Defaults to 100.
   */
  recursionLimit: number;
```

The default of 100 matches C++, Java, and Python's default.
When a message nests deeper than the limit, parsing fails with a catchable error before overflowing the stack:

```ts
try {
  fromBinary(schema, bytes);
} catch (e) {
  e; // cannot decode <type> from binary: maximum recursion depth of 100 reached
}
```

Note that this is a behavior change: input nested deeper than 100 levels used to decode and now throws. Code that relies on deeper nesting can raise the limit explicitly, e.g. `fromBinary(schema, bytes, { recursionLimit: 1000 })`.

The new option is made available on `getExtension` too, which didn't expose binary options so far.

### CODE REVIEW

**Review by ajeetdsouza — APPROVED** [review]

File: packages/protobuf-test/src/json.test.ts [hunk 1]
```diff
@@ -1294,6 +1294,76 @@ void suite("JSON parse errors", () => {
...
+    for (let i = 0; i < depth; i++) {
+      message = { recursiveMessage: message };
+    }
+    return message;
+  }
+
+  test("rejects deeply nested Any instead of overflowing the stack", () => {
+    const registry = createRegistry(AnySchema);
+    assert.throws(
+      () => fromJson(AnySchema, makeNestedAny(1000), { registry }),
```
Comment (ajeetdsouza) [inline packages/protobuf-test/src/json.test.ts:1325]: I would add a test on the boundary of the recursion limit to check for off-by-one errors.
